storage/innobase/buf/buf0buf.cc

```diff
@@ -1552,720 +1615,520 @@ buf_tmp_buffer_t *buf_pool_t::io_buf_t::reserve(bool wait_for_reads) noexcept
+    return SHRINK_ABORT;
+  }
 
-#ifdef BTR_CUR_HASH_ADAPT
-	/* enable AHI if needed */
-	if (btr_search_disabled) {
-		btr_search_enable(true);
-		ib::info() << "Re-enabled adaptive hash index.";
-	}
-#endif /* BTR_CUR_HASH_ADAPT */
+  try_LRU_scan= false;
+  mysql_mutex_unlock(&mutex);
+  page_cleaner_wakeup(true);
+  my_cond_wait(&done_flush_list, &flush_list_mutex.m_mutex);
+  mysql_mutex_unlock(&flush_list_mutex);
+  mysql_mutex_lock(&mutex);
 
-	if (warning)
-		buf_resize_status("Resizing buffer pool failed");
+  if (!n_blocks_to_withdraw)
+    goto withdraw_done;
 
-	ut_d(validate());
+  return SHRINK_IN_PROGRESS;
+}
 
-	return;
+inline void buf_pool_t::shrunk(size_t size, size_t reduced) noexcept
+{
+  ut_ad(size + reduced == size_in_bytes);
+  size_in_bytes_requested= size;
+  size_in_bytes= size;
+# ifdef __linux__
+  /* page_guess() may read this memory */
+  MEM_UNDEFINED(memory + size, size_in_bytes_max - size);
+# else
+  for (size_t n= page_hash.pad(page_hash.n_cells), i= 0; i < n;
+       i+= page_hash.ELEMENTS_PER_LATCH + 1)
+  {
+    auto &latch= reinterpret_cast<page_hash_latch&>(page_hash.array[i]);
+    latch.lock();
+    /* We already shrunk size_in_bytes. The exclusive lock here
+    ensures that any page_guess() will detect an out-of-bounds
+    guess before we invoke my_virtual_mem_decommit() below. */
+    latch.unlock();
+  }
+# endif
+  my_virtual_mem_decommit(memory + size, reduced);
+#ifdef UNIV_PFS_MEMORY
+  PSI_MEMORY_CALL(memory_free)(mem_key_buf_buf_pool, reduced, owner);
+#endif
 }
 
-#ifdef __linux__
-inline void buf_pool_t::garbage_collect()
+ATTRIBUTE_COLD void buf_pool_t::resize(size_t size, THD *thd) noexcept
 {
+  ut_ad(this == &buf_pool);
+  mysql_mutex_assert_owner(&LOCK_global_system_variables);
+  ut_ad(size <= size_in_bytes_max);
+  if (my_use_large_pages)
+  {
+    my_error(ER_VARIABLE_IS_READONLY, MYF(0), "InnoDB",
+             "innodb_buffer_pool_size", "large_pages=0");
+    return;
+  }
+
+  size_t n_blocks_new= get_n_blocks(size);
+
   mysql_mutex_lock(&mutex);
-  size_t freed= 0;
 
-#ifdef BTR_CUR_HASH_ADAPT
-  /* buf_LRU_free_page() will temporarily release and reacquire
-  buf_pool.mutex for invoking btr_search_drop_page_hash_index(). Thus,
-  we must protect ourselves with the hazard pointer. */
-rescan:
-#else
-  lru_hp.set(nullptr);
-#endif
-  for (buf_page_t *bpage= UT_LIST_GET_LAST(LRU), *prev; bpage; bpage= prev)
+  const size_t old_size= size_in_bytes;
+  if (first_to_withdraw || old_size != size_in_bytes_requested)
   {
-    prev= UT_LIST_GET_PREV(LRU, bpage);
-#ifdef BTR_CUR_HASH_ADAPT
-    lru_hp.set(prev);
+    mysql_mutex_unlock(&mutex);
+    my_printf_error(ER_WRONG_USAGE,
+                    "innodb_buffer_pool_size change is already in progress",
+                    MYF(0));
+    return;
+  }
+
+  ut_ad(UT_LIST_GET_LEN(withdrawn) == 0);
+  ut_ad(n_blocks_to_withdraw == 0);
+#ifdef __linux__
+  DBUG_EXECUTE_IF("trigger_garbage_collection",
```

Comment:
Please test garbage collection more explicitly. using suggested variable to trigger it. Not just on Linux.

Reply:
This is a comment that you should have posted in #2805. We are only adapting and improving the existing Linux specific interface here.

Reply:
It can easily become not-Linux only. I'm objecting to creation of Linux specific variables, and missing testability.

Reply:
84f1aeaccd168a592a804f7d6695a383ef3c55b7 should enable the test `innodb.mem_pressure` on all debug instrumented builds.